pkg/sql/schemachanger/doc.go

```diff
@@ -0,0 +1,169 @@
+//
+// We model these state transitions as a directed acyclical graph, in which each
+// node is an (element, target state, current state) tuple and state transitions
+// are edges between nodes sharing the same (element, target state) pairs. We
+// denote these as op-edges because these edges are decorated with the sequence
+// of operations to effect the state transition. Constraints on state
+// transitions, such as the 2-version invariant, are modeled as dependency
+// edges, or dep-edges, in the graph. These edges are decorated with constraints
+// which need to be satisfied. In the case of the 2-version invariant, the
+// constraint is `PreviousTransactionPrecedence`, which imposes that the state
+// transitions must reach the origin node before the destination node and more
+// importantly must do so in separate transactions.
+//
+// Planning a schema change therefore boils down to generating the corresponding
+// DAG and identifying a graph partition which satisfies all dep-edge
+// constraints. The graph object itself is defined in `scplan/internal/scgraph`,
+// the nodes and the op-edges are generated using `scplan/internal/opgen`, the
+// dep-edges between those nodes are generated using
+// `scplan/internal/rules/...`, and the partitioning is implemented in
+// `scplan/internal/scstage`. Unsurprisingly, the `scplan` package ties all of
+// this together.
+//
+// # Rules
+//
+// Generating these dep-edges is non-trivial and, practically speaking, requires
+// a vast quantity of complex code. This deserves further commentary.
+//
+// A graph will typically have a lot of dep-edges. Think for instance of a DROP
+// DATABASE statement operating on a database with a good many tables in it,
+// each table will easily have a few dozen elements and in such a DROP these
+// will all have target ABSENT. There will be dep-edges to ensure that, for
+// instance, a table name is not removed from system. namespace before the table
+// descriptor's `state` field is set to `DROP`.
+//
+// These edges need to be generated programmatically, and rather than doing so
+// in an imperative fashion, we have chosen to define them declaratively, in the
+// form of rules. These rules are essentially predicates which can be applied to
+// any pair of nodes to determine whether there should be a dep-edge between
+// those to. See `scplan/internal/rules/current` for the latest set of rules.
+//
+// The advantage of this approach is that the rules can be expressed
+// using a high level of abstraction, making them easier to reason about and
+// therefore less likely to have bugs. The disadvantage is that we needed to
+// build what's essentially a mini-embedded RDBMS to apply these predicates
+// efficiently, from a performance perspective. This database is implemented in
+// `rel`.
+//
+// This is the reason why the element model is relational in nature. Each
+// element can be keyed (the keys are defined in `screl`) and therefore joined
```

Comment:
nit: maybe say this instead: Each element can be keyed by their attributes (the attributes are defined in `screl/attr.go`